components/nimbus/src/enrollment.rs

```diff
@@ -601,30 +614,116 @@ impl<'a> EnrollmentsEvolver<'a> {
                 is_user_participating,
                 existing_experiments.get(slug).copied(),
                 updated_experiments.get(slug).copied(),
-                existing_enrollments.get(slug).copied(),
+                existing_enrollment_map.get(slug).copied(),
                 &mut enrollment_events,
             )?;
-            if let Some(enrollment) = updated_enrollment {
-                updated_enrollments.push(enrollment);
+            // if let Some(_enrollment) = updated_enrollment {
+            // What has changed? If we're still enrolled, do nothing.
+
+            // If not, remove the feature id - it's usable now.
+
+            // updated_enrollments.push(enrollment); XXX NOT NEEDED HERE, I
+            //if enrollment.status != EnrollmentStatus::Enrolled
+            // }
+        }
+
+        // Step 3: now we can add new experiments that avoid the features that
+        // already being experimented upon.
+
+        // the previous calls to evolve_enrollments will have prepopulated
+        // so...
+        enrollment_events.clear();
+
+        for slug in all_slugs {
+            let updated_experiment = updated_experiments.get(slug).copied();
+            if let Some(unwrapped_experiment) = updated_experiment {
+                let feature_id = &unwrapped_experiment.get_first_feature_id();
+
+                // If this feature_id is locally free, evolve the enrollment
+                // update the feature_id hashtable
+                if locally_enrolled_feature_ids.get(feature_id) == None {
+                    log::debug!("this feature is locally unused");
+
+                    let updated_enrollment = self.evolve_enrollment(
+                        is_user_participating,
+                        existing_experiments.get(slug).copied(),
+                        updated_experiments.get(slug).copied(),
+                        existing_enrollment_map.get(slug).copied(),
+                        &mut enrollment_events,
+                    )?;
+
+                    if let Some(enrollment) = updated_enrollment {
+                        log::debug!(
+                            "updated enrollment was returned from evolve_enrollment; pushing"
+                        );
+                        updated_enrollments.push(enrollment);
+
+                        log::debug!(
+                            "mapping {}:{} to locally_enrolled_feature_ids",
+                            feature_id,
+                            slug
+                        );
+                        locally_enrolled_feature_ids.insert(feature_id.clone(), slug.clone());
+                    }
+                } else {
+                    // Does this experiment have a feature id that we've already used?
+                    log::debug!(
+                        "feature_id {} already used by experiment, marking {} NotEnrolled",
+                        feature_id,
+                        slug
+                    );
```

Comment:
Here, we only know that the `feature_id` was used by _an_ experiment. It might have been used by _this_ experiment.

We should check either `locally_enrolled_feature_ids.get(feature_id) == Some(slug)` or `existing_enrollment_map.get(slug).is_some()` before registering a non_enrollment.

Reply:
For clarity, I think I prefer the former.